### ISSUE TITLE
Wrong image dimensions when cropping image with a reference image

### ISSUE THREAD

**Issue by andreanne-lemay** [neuropoly/spinalcordtoolbox#2768]
### Description
When an image (with the same physical resolution as the reference image) is cropped with the command-line `sct_crop_image -ref`, the resulting cropped image doesn't have the same dimensions as the reference image. There is a 1 pixel difference in some dimensions. I suspect some sort of precision error. This occurs on the spinal cord tumor dataset, but this error is not present on all datasets.

### Steps to Reproduce

Run the following lines on at this location: `duke/temp/andreanne/tumor_segmentation_new_edema/results/data/sub-Astr158/anat`
```
python
from spinalcordtoolbox.image import Image
import os
os.system("sct_resample -i sub-Astr158_T2w.nii.gz -o sub-Astr158_T2w_resampled.nii.gz -mm 0.5x0.5x2")
os.system("sct_resample -i sub-Astr158_T1w.nii.gz -o sub-Astr158_T1w_resampled.nii.gz -mm 0.5x0.5x2")

t2w = Image("sub-Astr158_T2w_resampled.nii.gz")
t1w = Image("sub-Astr158_T1w_resampled.nii.gz")

t2w.dim # the following dimensions should be displayed: (560, 560, 18, 1, 0.5, 0.5, 2.0166671, 1)
t1w.dim # the following dimensions should be displayed: (600, 600, 18, 1, 0.5, 0.5, 2.0166671, 1)

os.system("sct_crop_image -i sub-Astr158_T1w_resampled.nii.gz -ref sub-Astr158_T2w_resampled.nii.gz -o sub-Astr158_T1w_resampled_crop.nii.gz")

t1w_crop = Image("sub-Astr158_T1w_resampled_crop.nii.gz")
t1w_crop.dim # the following dimensions should be displayed: (560, 559, 17, 1, 0.5, 0.5, 2.0166671, 1)
t2w.dim # the following dimensions should be displayed: (560, 560, 18, 1, 0.5, 0.5, 2.0166671, 1)
```
**Expected behavior:** The cropped image and the reference image should have the same dimensions.

**Actual behavior:** The cropped image doesn't have the same dimensions as the reference image.

![image](https://user-images.githubusercontent.com/49137243/86390371-ef41b100-bc65-11ea-9b46-f89e1ee99c47.png)

**Comment by cfhammill** [comment]
I can reproduce this issue

```
sct_download_data -d course_london19
cd course_london19/single_subject/data/t1/ 
sct_resample -i t1.nii.gz -o t1_small.nii.gz -mm 2.2x1.0x1.0
sct_resample -i t1.nii.gz -o t1_smaller.nii.gz -mm 2.4x1.0x1.0
sct_crop_image -i t1_small.nii.gz -o t1_small_crop.nii.gz -ref t1_smaller.nii.gz
```

t1_small.nii.gz (87, 260, 320)
t1_small_crop.nii.gz (85, 260, 319)

**Comment by cfhammill** [comment]
So the `sct_crop_image` creates an `ImageCropper` object which has method `crop` which calls method `get_bbox_from_ref` when a reference image is supplied. The docstring for `get_box_from_ref` says `"Get bounding box from input reference image, by looking at min/max indices in each dimension. img_ref and self.img_in should have the same dimensions."`. So it seems like this code wasn't designed for this use case. 

What goes wrong is that the reference image is filled with 1s, and then resampled to the input image using `nib_resample` with nearest neighbour interpolation. Then a bounding box is computed for the resampled image which either by a resampling artifact or a computational error is losing a voxel on the boundary.

**Comment by jcohenadad** [comment]
@andreanne-lemay if i understand correctly, what you would like to do is bring image A into the voxel-coordinate space of image B. If that's what you want to do, then i would instead use:
~~~
sct_register_multimodal -i A -d B -identity 1 -x <YOUR-FAVORITE-INTERPOLATION-METHOD>
~~~

**Comment by jcohenadad** [comment]
regardless, we should clarify the documentation for `sct_crop_image -ref`

**Comment by andreanne-lemay** [comment]
> regardless, we should clarify the documentation for `sct_crop_image -ref`, i.e. `-ref` is supposed to be a binary image, with an object (e.g. a box), that defines the bounding box for the cropping of `-i`.

I thought the cropping you are describing was with the option -m `Binary mask that will be used to extract bounding box for cropping the image. Has priority over -ref.` .

**Comment by jcohenadad** [comment]
> I thought the cropping you are describing was with the option -m Binary mask that will be used to extract bounding box for cropping the image. Has priority over -ref. .

ha! you're right 👍 

ok, so now i am wondering what is the use-case of this function. Have you tried https://github.com/neuropoly/spinalcordtoolbox/issues/2768#issuecomment-656873538 and does it achieve what you are trying to do?

**Comment by andreanne-lemay** [comment]
> ok, so now i am wondering what is the use-case of this function. Have you tried [#2768 (comment)](https://github.com/neuropoly/spinalcordtoolbox/issues/2768#issuecomment-656873538) and does it achieve what you are trying to do?
> 
> if it does, then we should re-visit the purpose of -ref

Yes, it worked! As a user, it is more intuitive to go look in sct_crop_image for cropping an image compared to sct_register_multimodal, so I believe there is a purpose to having this method in sct_crop_image. Maybe as a quick fix, the code from `sct_register_multimodal -identity 1` could be reused for `sct_crop_image -ref`?

**Comment by jcohenadad** [comment]
> Maybe as a quick fix, the code from `sct_register_multimodal -identity 1` could be reused for `sct_crop_image -ref`?

yes, this sounds like a reasonable approach. For now you have a workaround so this is not top priority.

**Comment by jcohenadad** [comment]
added priority flag because the change proposed here will break compatibility with previous CLI

**Comment by Drulex** [comment]
I thought I would be able to quickly fix this, however looking at `sct_register_multimodal.py` it appears that the code that handles the identity option is heavily coupled with the whole registration logic making it not trivial to re-use.

What I propose is that we implement the feature in the appropriate API module (seems to be `sct.cropping` from what I gather in this thread). We can then import the functionality in `sct_cropping` and later `sct_register_multimodal` when it is properly refactored (if still applicable).

What do you think @jcohenadad ?

**Comment by andreanne-lemay** [comment]
I confirm that this issue has not been solved (I can still reproduce the error with the images in `duke/sct_testing/beijing_tumor/sub-Astr158/anat`).

**Comment by dpapp86** [comment]
Just pinging that this is still an issue at time of writing, see this thread:

The suggested workaround works, but as a (not so) naive user, I would have used the -ref option.

https://forum.spinalcordmri.org/t/sct-crop-image-ref-result-is-one-pixel-smaller/1116/
